### ISSUE TITLE
VIM: Command for navigating around visible editors/viewlets/repl

### ISSUE THREAD

**Issue by rebornix** [Microsoft/vscode#21669]
- VSCode Version: 1.10
- OS Version: macOS

Trying to adopt navigation commands in list, they work pretty well and it doesn't require any code change in Vim extensions. Right now the only missing part is how we navigate to the visible viewlet or repl.

For split editors, right now the way to workaround is check the column of current editor, and try to switch to previous/next one when users press `<C-w> hjkl`. But users have no way to navigate to other part of VS Code by a generic command and we have no idea which component is visible from extension host.

A proper command can be something like

```
{
    "command": "navigate",
    "args": "up" // "down", "left"
}
```

It's similar to Tab in Accessibility Mode  or `workbench.action.navigateEditorGroups`, users can use limited commands to switch focus between visible areas in the workspace.

**Comment by bpasero** [comment]
@rebornix to be clear we are talking about navigating these major parts of the VS Code UI, specifically:
* the view in the side bar
* the panel
* up to three editors

Is there anything more to consider? What happens if you "navigate down" when you are already in the panel?

**Comment by bpasero** [comment]
Also, what does it mean when editors are layout horizontally, would you then expect Arrow-Up/Down to navigate the editors and Arrow-Left/Right not?

**Comment by bpasero** [comment]
Marking as help wanted, this would be a good area for a PR.

**Comment by misoguy** [comment]
I'd like to give this a try if no one is on it yet :)

**Comment by rebornix** [comment]
@bpasero my 2 cents to your questions

* The command can be used to target visible parts, including the view in the side bar, the panel and editors. IMHO these three UI components are enough.
* We only navigate between visible parts, so if the part is hidden, we can't navigate to it. Besides, we never open any part for users, that feature belongs to our `Toggle xyz` commands.
* About navigating down if you are in the panel, I personal preference is doing nothing in this case. Others might prefer to loop though all the visible parts in a direction but it will be too complicated.
* Since we are navigating through Visible parts, we should use Arrow keys as they are. When editors are layout horizontally, we should use Arrow-Up/Down.

This command only focuses on components that are already visible in the viewport and don't really care what they are. To some extent, it's a combination of Focus xyz Editor Groups/Focus Panel/Focus Sidebar, with a simple key cord `<modifier>+arrows`.

**Comment by rebornix** [comment]
@misoguy go ahead :)

**Comment by bpasero** [comment]
@rebornix thanks man! @misoguy yeah, go ahead and I can review, thanks 👍

**Comment by misoguy** [comment]
I'm currently digging around to see how to implement this and have a few questions.

1. Where should I implement these actions? Because what i have understood so far is that each view(the viewlet, editor and panel) has its own actions implemented on their relative files.

2. Should each navigating action be implemented seperately? ex:
    - ```workbench.action.navigateLeft```
    - ```workbench.action.navigateRight```
    - ```workbench.action.navigateUp```
    - ```workbench.action.navigateDown```

3. If implemented, should these actions be mapped to a key ```<modifier>+arrows``` by default?

**Comment by bpasero** [comment]
@misoguy I am fine to add these into https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/electron-browser/actions.ts#L512 and I agree we need one command for each, but I think we can have a base action that the others extend from to not duplicate code too much.

I think we do not need to map them to any keybinding out of the box, but we can always tweak that later.

As for how to navigate between views, I suggest you look how our view picker is able to navigate to views and panels: https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/parts/quickopen/browser/viewPickerHandler.ts#L1

There is also an `IEditorGroupService` that can be used to see how many editors are open ("editor stacks model") and to navigate between groups.

**Comment by misoguy** [comment]
Thanks for the quick and detail explaination! I'll try to get on it as you said :)

**Comment by misoguy** [comment]
This feature seems like there is more work than I anticipated.
I bumped into a question while working on it.

Is there an API to check if the current active editor is the first or last in a group? Something like
```isFirstEditorOfGroup``` and ```isLastEditorOfGroup```
If not, would it be better to add it in the ```interface IEditorInput``` or just implement it in the base action just for this action? (Its required for case when i want to jump from the first editor in the first group to the sidebar if the sidebar is at the left and vice versa when sidebar is at the right)

**Comment by misoguy** [comment]
After looking more into the code, I got the feeling that instead of constantly calculating if current active editor is the first or last in a group, it is better to add an optional argument into ```interface IEditorStacksModel``` to stop cycling to the first/last group. I'll add a PR to explain this in code.

**Comment by bpasero** [comment]
@misoguy @rebornix I think we do not want to cycle if at the end, or did I get that wrong in any way?

@misoguy btw you have to accommodate for the fact that the sidebar can be positioned to the right via the setting `"workbench.sideBar.location: right"`. 

The other setting to keep in mind is vertical vs horizontal orientation of editors via [`IEditorGroupService.getGroupOrientation()`](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/services/group/common/groupService.ts#L93)

And finally as for the position of the editor, you can just use [`IEditorStackModel.positionOfGroup()`](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/common/editor.ts#L842) to convert a group into a number (Position is an enum). 0 will be the first, 1 the second and 2 the third.

**Comment by misoguy** [comment]
@bpasero Currently the default behavior of ```next()``` and ```previous()``` (implemented with keybinding cmd+alt+left/right arrow)is to cycle when at the end/start of the stack if I'm not wrong, hence the option to stop cycling. 

And as for the sidebar location being able to be at the right of the screen is already taken into account and the current setting can be read from [```IPartService.getSideBarPosition()```](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/services/part/common/partService.ts#L103).

The vertical and horizontal orientation will also be considered when implementing. Thanks for the heads up about the function :)

**Comment by bpasero** [comment]
Ah yeah good to have an option for `next/previous` then on the stacks model (can also be tested). 

GroupOrientation pretty much is an enum with this declaration: 

`export type GroupOrientation = 'vertical' | 'horizontal';`

Just not a number enum.

**Comment by misoguy** [comment]
Sorry to ask so much but I got one more question while implementing.
If my focus is at the sidebar(on the left) when both the editor and panel is open, and if I were to navigate to the right, should I focus on the first editor or the panel?
I personality think that going to the editor is the right move but I'm asking just to be sure we are on the same page. :)

**Comment by rebornix** [comment]
Left to right and top to bottom is my preference, which is going to the
editor instead of panel.

**Comment by misoguy** [comment]
I've implemented the navigation on the PR #22005, it'd be nice to get some comments on it :)

**Comment by cleidigh** [comment]
@misoguy 
just checked out and tried out your code , looks good nice addition for us keyboard users
I'm working on a keyboard expand contract for visible parts may have to borrow from your similar stuff ;-)
Hope your review mine as well !

**Comment by misoguy** [comment]
@cleidigh I'll gladly review when you have a PR ready! :)

**Comment by bpasero** [comment]
Verify:
* new commands `workbench.action.navigateDown|Up|Left|Right` can be used to navigate across all visible views (sidebar, panel, editors) in the order of their appearance
* verify with sidebar left and right
* verify with hidden sidebar and panel and multiple editor groups

**Comment by theduke** [comment]
Awesome.
